primitives/io/src/lib.rs

```diff
@@ -166,11 +188,53 @@ pub trait Storage {
 	/// Use this function to distribute the deletion of a single child trie across multiple
 	/// blocks.
 	#[version(2)]
+	fn clear_prefix(&mut self, prefix: &[u8], limit: Option<u32>) -> KillStorageResultPrev {
+		let clear_prefix_result = Externalities::clear_prefix(*self, prefix, limit);
+		match clear_prefix_result.all_keys_removed {
+			true => KillStorageResultPrev::AllRemoved(clear_prefix_result.num_keys_from_backend),
+			false =>
+				KillStorageResultPrev::SomeRemaining(clear_prefix_result.num_keys_from_backend),
+		}
+	}
+
+	/// Clear the storage of each key-value pair where the key starts with the given `prefix`.
+	///
+	/// # Limit
+	///
+	/// Deletes all keys from the overlay and up to `limit` keys from the backend if
+	/// it is set to `Some`. No limit is applied when `limit` is set to `None`.
+	///
+	/// The limit can be used to partially delete a prefix storage in case it is too large
+	/// to delete in one go (block).
+	///
+	/// Also returns a struct `KillStorageResult` with
+	/// the number of keys removed from the backend and the overlay.
+	///
+	/// # Note
+	///
+	/// Please note that keys that are residing in the overlay for that prefix when
+	/// issuing this call are all deleted without counting towards the `limit`. Only keys
+	/// written during the current block are part of the overlay. Deleting with a `limit`
+	/// mostly makes sense with an empty overlay for that prefix.
+	///
+	/// Calling this function multiple times per block for the same `prefix` does
+	/// not make much sense because it is not cumulative when called inside the same block.
+	/// Use this function to distribute the deletion of a single child trie across multiple
+	/// blocks.
+	#[version(3)]
 	fn clear_prefix(&mut self, prefix: &[u8], limit: Option<u32>) -> KillStorageResult {
-		let (all_removed, num_removed) = Externalities::clear_prefix(*self, prefix, limit);
-		match all_removed {
-			true => KillStorageResult::AllRemoved(num_removed),
-			false => KillStorageResult::SomeRemaining(num_removed),
+		let clear_prefix_result = Externalities::clear_prefix(*self, prefix, limit);
+
+		if clear_prefix_result.all_keys_removed {
+			KillStorageResult::AllRemoved {
+				backend: clear_prefix_result.num_keys_from_backend,
+				overlay: clear_prefix_result.num_keys_from_overlay.unwrap(),
```

Comment:
we shouldn't need to unwrap here, or maybe `unwrap_or(0)`, but actually I don't really understand why ClearPrefixResult use an `Option<u32>` instead of just a u32. What does None means ? is it different from 0 ?

Reply:
There are some functions that returns ClearPrefixResult with no numbers of overlay keys in them. That's why I made the number of  overlay keys an option

Reply:
you basic externalities ? so the unwrap would fail here in this case no ?
In basic externalities we consider that there is no overlay, I think it is better to just return that 0 key were removed from overlay and make the field a simple `u32`

Reply:
Oh... That's what I implemented earlier but I probably over thought that's why I did this. 
I will change the option to just u32 then.
Thanks